From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gr1mly4Memes <alagozantoni14@gmail.com>
Date: Sat, 21 Feb 2026 08:13:16 -0800
Subject: [PATCH] Add config


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 26a9fd255156a24946bf79eaf6b686c27e324b88..713b62403304a031c948e2b9d343e5292fdf96e0 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1162,6 +1162,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper end - rewrite chunk system
         // Paper start - Improved watchdog support - move final shutdown items here
         Util.shutdownExecutors();
+        gr1mly4memes.papercut.PapercutConfig.createWorldSections = false; // Papercut
         this.onServerExit();
         // Paper end - Improved watchdog support - move final shutdown items here
     }
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 3d8290b952e4341211ef8e669372bf7ea1c4bf62..172cbe6ff82836ca2041d3c9db32ffe96b7a994c 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -274,6 +274,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.server.spark.registerCommandBeforePlugins(this.server); // Paper - spark
         com.destroystokyo.paper.Metrics.PaperMetrics.startMetrics(); // Paper - start metrics
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
+        gr1mly4memes.papercut.PapercutConfig.init((java.io.File) options.valueOf("papercut-settings")); // Papercut - Server Config
 
         // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
