From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gr1mly4Memes <alagozantoni14@gmail.com>
Date: Sat, 28 Feb 2026 20:48:05 -0800
Subject: [PATCH] Fix Enderpearl get removed when the chunk unload


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index 3fbd4f7e563bcb8da86c66bdb9b57ae0f7bda397..98c629e00ec603440e45f940914a74fbf02be2d5 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -258,7 +258,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
             int sectionPosZ = SectionPos.blockToSectionCoord(this.position().z());
             Entity entity = this.owner != null ? findOwnerIncludingDeadPlayer(serverLevel, this.owner.getUUID()) : null;
             if (entity instanceof ServerPlayer serverPlayer
-                && !entity.isAlive()
+                && serverPlayer.getHealth() <= 0.0f // Papercut - fix Enderpearl get removed when the chunk unload
                 && !serverPlayer.wonGame
                 && serverPlayer.level().getGameRules().get(GameRules.ENDER_PEARLS_VANISH_ON_DEATH)) {
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
