From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gr1mly4Memes <alagozantoni14@gmail.com>
Date: Sun, 1 Mar 2026 10:48:25 -0800
Subject: [PATCH] Disable Packet Limit


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 8835cf8d9d1fb3b4930a50d002e0b426b155f4ac..e6a1ab10d4b63c9bb51bd50d33e7c6dfc6f2394e 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -272,7 +272,7 @@ public class GlobalConfiguration extends ConfigurationPart {
             }
 
             public boolean isEnabled() {
-                return this.interval > 0.0 && this.maxPacketRate > 0.0;
+                return !gr1mly4memes.papercut.PapercutConfig.disablePacketLimit && (this.interval > 0.0 && this.maxPacketRate > 0.0); // Papercut
             }
 
             public enum ViolateAction {
